File: solutions/5kyu/product_fib.c
/*
*   kata link
*   https://www.codewars.com/kata/5541f58a944b85ce6d00006a
* */

#include<stdlib.h>
#include<stdio.h>

typedef unsigned long long ull;

unsigned long long* product_fib(ull prod) {
  ull a = 1 , b = 1 ;
  while(a * b < prod){
    b = a + b ;
    a = b - a ;
  }

  ull * res ;
  res = (ull *) malloc (3*sizeof(ull));
  res[0] = a ;
  res[1] = (ull )b;
  res[2] = (ull)(a*b == prod) ;
  return res;
}

File: solutions/5kyu/ProductFib.java
/*
*   kata link
*   https://www.codewars.com/kata/5541f58a944b85ce6d00006a
*
*  */
public class ProdFib { // must be public for codewars	

    public static long[] productFib(long prod) {
        long f1 = 0, f2 = 1;
        while(f1 * f2 < prod){
            f2 = f1 + f2;
            f1 = f2 - f1;
        }
        return f1 * f2 == prod? new long[] {f1, f2 , 1} : new long[]{f1, f2, 0};
    }
}
